apex/contrib/optimizers/distributed_fused_adam.py

```diff
@@ -1794,29 +1794,66 @@ def step(self, closure=None, *, grad_scaler=None):
                     device=self.device,
                 )
 
-        # Apply optimizer step
+        # Apply optimizer step and synchronize params
         self.state['step'] += 1
-        if self.store_param_remainders:
-            self._local_step_with_param_remainders()
-        else:
-            self._local_step()
+        if self.distributed_size > 1 and self.overlap_param_sync and self.state['buckets']:
+            # Local step and non-blocking param sync
+            # Note: Overlap param sync of first buckets with optimizer
+            # step of remaining buckets.
+
+            # Get buckets containing "first" parameter
+            fragment = self.state['buckets'][-1].fragments[-1]
```

Comment:
excuse me for my lack of understanding but why the last bucket and the last fragment?

Reply:
By default, each param initializes its optimizer state in the first backward pass, so the first param to show up in the forward pass should be the last one to initialize.